appservice/consumers/roomserver.go

```diff
@@ -151,15 +152,25 @@ func (s *OutputRoomEventConsumer) onMessage(
 		return true
 	}
 
+	txnID := ""
+	// Try to get the message metadata, if we're able to, use the timestamp as the txnID
+	if len(msgs) > 0 {
+		metadata, err := msgs[0].Metadata()
+		if err == nil {
+			txnID = strconv.Itoa(int(metadata.Timestamp.UnixNano()))
+		}
+	}
+
 	// Send event to any relevant application services. If we hit
 	// an error here, return false, so that we negatively ack.
 	log.WithField("appservice", state.ID).Debugf("Appservice worker sending %d events(s) from roomserver", len(events))
-	return s.sendEvents(ctx, state, events) == nil
+	return s.sendEvents(txnID, ctx, state, events) == nil
 }
 
 // sendEvents passes events to the appservice by using the transactions
 // endpoint. It will block for the backoff period if necessary.
 func (s *OutputRoomEventConsumer) sendEvents(
+	txnID string,
```

Comment:
Can you move the new parameter to the end?
Having `context.Context` as the first parameter is more idiomatic Go.

Reply:
done.